src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,575 @@
+                         discount=discount,
+                         gae_lambda=gae_lambda,
+                         center_adv=center_adv,
+                         positive_adv=positive_adv,
+                         fixed_horizon=fixed_horizon,
+                         flatten_input=flatten_input)
+
+    @overrides
+    def init_opt(self):
+        """Initialize optimizater."""
+        pol_loss_inputs, pol_opt_inputs = self._build_inputs()
+        self._policy_opt_inputs = pol_opt_inputs
+
+        pol_loss, pol_kl = self._build_policy_loss(pol_loss_inputs)
+        self.optimizer.update_opt(loss=pol_loss,
+                                  target=self.policy,
+                                  leq_constraint=(pol_kl, self.max_kl_step),
+                                  inputs=flatten_inputs(
+                                      self._policy_opt_inputs),
+                                  constraint_name='mean_kl')
+
+        return dict()
+
+    @overrides
+    def optimize_policy(self, itr, samples_data):
+        """Optimize policy."""
+        policy_opt_input_values = self._policy_opt_input_values(samples_data)
+        # Train policy network
+        logger.log('Computing loss before')
+        loss_before = self.optimizer.loss(policy_opt_input_values)
+        logger.log('Computing KL before')
+        policy_kl_before = self.f_policy_kl(*policy_opt_input_values)
+        logger.log('Optimizing')
+        self.optimizer.optimize(policy_opt_input_values)
+        logger.log('Computing KL after')
+        policy_kl = self.f_policy_kl(*policy_opt_input_values)
+        logger.log('Computing loss after')
+        loss_after = self.optimizer.loss(policy_opt_input_values)
+        tabular.record('{}/LossBefore'.format(self.policy.name), loss_before)
+        tabular.record('{}/LossAfter'.format(self.policy.name), loss_after)
+        tabular.record('{}/dLoss'.format(self.policy.name),
+                       loss_before - loss_after)
+        tabular.record('{}/KLBefore'.format(self.policy.name),
+                       policy_kl_before)
+        tabular.record('{}/KL'.format(self.policy.name), policy_kl)
+        pol_ent = self.f_policy_entropy(*policy_opt_input_values)
+        tabular.record('{}/Entropy'.format(self.policy.name), np.mean(pol_ent))
+
+        self._fit_baseline(samples_data)
+        self._policy_update_ops()
+
+    @overrides
+    def get_itr_snapshot(self, itr):
+        """Get iteration snapshot."""
+        return dict(
+            itr=itr,
+            policy=self.policy,
+            baseline=self.baseline,
+        )
+
+    def _build_inputs(self):
+        action_space = self.policy.action_space
+
+        with tf.name_scope('inputs'):
+            obs_var = self.policy.model.networks['default'].input
+            old_obs_var = self.old_policy.model.networks['default'].input
+            # Ops to assign new policy weights to old policy
+            target_update_op = tensor_utils.get_target_ops(
```

Comment:
hmm can you explain using this vs `self._old_policy.parameters = self._policy.parameters`? I am genuinely curious.

Reply:
When assigning parameters, the `model` class matches the variables by full name - it only works when pickling the same object. Unless we change the "name checking" methodology (e.g. match the name excluding primitive itself), we still have to manually assign the vars.

Reply:
also since policy is not yet a `model`, we need to do `self._old_policy.model.parameters = self._policy.model.parameters`.

Reply:
is there a way we can do this without using an op? can we add or change something in `model`? what are the risks of removing the full name checking?

one of the points of `model` is to avoid this thing in tensorflow where i have to make an op for things which aren't feed-forward.

we really want an API which is like `self._old_policy.model.parameters = self._policy.model.parameters`. i think it is more important to make this use case nice than the pickling one (pickling is already nasty).